odiglet/pkg/ebpf/manager.go

```diff
@@ -0,0 +1,371 @@
+}
+
+func NewManager(client client.Client, logger logr.Logger, factories map[FactoryID]Factory, procEvents chan detector.ProcessEvent) *Manager {
+	return &Manager{
+		ProcEvents:    procEvents,
+		Client:        client,
+		Factories:     factories,
+		Logger:        logger,
+		DetailsByPid:  make(map[int]InstrumentationDetails),
+		configUpdates: make(chan configUpdate),
+		done:          make(chan struct{}),
+	}
+}
+
+func (m *Manager) UpdateConfig(ctx context.Context, workloadKey types.NamespacedName, config *odigosv1.InstrumentationConfig) error {
+	// send a config update event for the specified workload
+	select {
+	case m.configUpdates <- configUpdate{workloadKey: workloadKey, config: config}:
+		return nil
+	case <-ctx.Done():
+		if ctx.Err() == context.DeadlineExceeded {
+			return errors.New("failed to update config of workload: timeout waiting for config update")
+		}
+		return ctx.Err()
+	}
+}
+
+func (m *Manager) Run(ctx context.Context) {
+	defer close(m.done)
+
+	runLoopCtx, stop := context.WithCancel(ctx)
+	m.stop = stop
+
+	// main event loop for handling instrumentations
+	for {
+		select {
+		case <-runLoopCtx.Done():
+			m.Logger.Info("stopping Odiglet instrumentation manager")
+			for pid, details := range m.DetailsByPid {
+				err := details.Inst.Close(ctx)
+				if err != nil {
+					m.Logger.Error(err, "failed to close instrumentation", "pid", pid)
+				}
+				delete(m.DetailsByPid, pid)
+				// probably shouldn't remove instrumentation instance here
+				// as this flow is happening when Odiglet is shutting down
+			}
+			return
+		case e := <-m.ProcEvents:
+			switch e.EventType {
+			case detector.ProcessExecEvent:
+				m.Logger.Info("detected new process", "pid", e.PID, "cmd", e.ExecDetails.CmdLine)
+				err := m.handleProcessExecEvent(runLoopCtx, e)
+				// ignore the error if no instrumentation factory is found,
+				// as this is expected for some language and sdk combinations
+				if err != nil && !errors.Is(err, ErrNoInstrumentationFactory) {
+					m.Logger.Error(err, "failed to handle process exec event")
+				}
+			case detector.ProcessExitEvent:
+				m.cleanInstrumentation(runLoopCtx, e.PID)
+			}
+		case configUpdate := <-m.configUpdates:
+			if configUpdate.config == nil {
+				m.Logger.Info("received nil config update, skipping")
+			}
+			for _, sdkConfig := range configUpdate.config.Spec.SdkConfigs {
+				err := m.applyInstrumentationConfigurationForSDK(runLoopCtx, configUpdate.workloadKey, &sdkConfig)
+				if err != nil {
+					m.Logger.Error(err, "failed to apply instrumentation configuration")
+				}
+			}
+		}
+	}
+}
+
+func (m *Manager) Stop() {
+	m.stop()
+	<-m.done
+}
+
+func (m *Manager) cleanInstrumentation(ctx context.Context, pid int) {
+	details, found := m.DetailsByPid[pid]
+	if !found {
+		m.Logger.V(3).Info("no instrumentation found for exiting pid", "pid", pid)
+		return
+	}
+
+	m.Logger.Info("cleaning instrumentation resources", "pid", pid)
+
+	err := details.Inst.Close(ctx)
```

Comment:
same nil check, maybe we should move the `Close` function to the details instance instead of closing the individual fields?
This will probably never be nil but still afraid of panics

Reply:
I created the `startTrackInstrumentation` which is the only place where `instrumentationDetails` is created, since this is an internal logic to this package, and not dependant on user input I think if the `inst` is nill - it is a result of some logical error on our side which should never happen.